### PR TITLE
feat(auto-balance): disk in pressure

### PR DESCRIPTION
**Pull request by c3y1huang**
#### Which issue(s) this PR fixes:

Issue longhorn/longhorn#4105

#### What this PR does / why we need it:

Implement a new function to sync data contents by local files.

#### Special notes for your reviewer:

`None`

#### Additional documentation or context

https://github.com/longhorn/longhorn-engine/pull/1003#pullrequestreview-2150095135

### CODE REVIEW

File: sparse/local.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	})
+
+	fileInfo, err := os.Stat(sourceFilePath)
+	if err != nil {
+		log.WithError(err).Error("Failed to get file info of source file")
+		return err
+	}
+
+	fileSize := fileInfo.Size()
+	if fileSize%Blocks != 0 {
```
Comment (derekbit) [inline sparse/local.go:28]: nit: `Blocks` is the block size here and a bit weird. We can rename it to block size in the future.

**Comment by derekbit** [comment]
Did a quick review.

What `syncContent()` does
1. Open the source file once and get a source file descriptor
2. Create the target file once and get a target file descriptor
3. Get data interval layout from the source file 
4. Initiate multiple workers (numSyncWorkers) for concurrently reading from the source file and writing to the target file.

We don't have any synchronization in the implementation. The concurrent readers and writers on a single source file descriptor and a single target file descriptor can indeed introduce race conditions.

**Comment by c3y1huang** [comment]
> Did a quick review.
> 
> What `syncContent()` does
> 
> 1. Open the source file once and get a source file descriptor
> 2. Create the target file once and get a target file descriptor
> 3. Get data interval layout from the source file
> 4. Initiate multiple workers (numSyncWorkers) for concurrently reading from the source file and writing to the target file.
> 
> We don't have any synchronization in the implementation. The concurrent readers and writers on a single source file descriptor and a single target file descriptor can indeed introduce race conditions.

Wouldn't they be on different interval due to the fileIntervalChannel?

**Comment by derekbit** [comment]
> Wouldn't they be on different interval due to the fileIntervalChannel?

Each worker seeks the file descriptor to a specific offset and reads or writes data, so multiple workers will race.

After discussing with @c3y1huang, we decided to use a single worker for now. We can improve it by using multiple workers in the future.

Using a single worker might not be a bad approach because:
- Sequential read and write operations usually perform well.
- We don't consume the entire disk IOPS.

cc @innobead

File: sparse/local.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+	}
+	defer sourceFileIo.Close()
+
+	local.sourceFileIo = sourceFileIo
+
+	targetFileIo, err := NewDirectFileIoProcessor(local.targetFilePath, os.O_RDWR|os.O_CREATE, 0)
+	if err != nil {
+		return err
+	}
+	local.targetFileIo = targetFileIo
```
Comment (derekbit) [inline sparse/local.go]: Can we open the two files inside `newSyncLocal()`? Then, they can be closed in `local.close()`.

File: sparse/local.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+func newSyncLocal(sourceFilePath, targetFilePath string, syncBatchSize int64) (*syncLocal, error) {
+	sourceFileIo, err := NewDirectFileIoProcessor(sourceFilePath, os.O_RDONLY, 0)
+	if err != nil {
+		return nil, err
+	}
+
+
+	targetFileIo, err := NewDirectFileIoProcessor(targetFilePath, os.O_RDWR|os.O_CREATE, 0)
+	if err != nil {
+		return nil, err
```
Comment (derekbit) [inline sparse/local.go:87]: close `sourceFileIo` if failed.

File: sparse/test/local_test.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+	. "github.com/longhorn/sparse-tools/sparse"
+)
+
+func TestSyncLocalFile(t *testing.T) {
+	layoutFrom := []FileInterval{
+		{Kind: SparseData, Interval: Interval{Begin: 0, End: 1 * Blocks}},
+		{Kind: SparseHole, Interval: Interval{Begin: 1 * Blocks, End: 2 * Blocks}},
+		{Kind: SparseData, Interval: Interval{Begin: 2 * Blocks, End: 3 * Blocks}},
+	}
```
Comment (derekbit) [inline sparse/test/local_test.go]: Can we add more test cases

1. data-hole-data (you've added)
2. hole-data-hole
3. hole-data
4. data-hole

Thank you.

**Review by derekbit — APPROVED** [review]

**Pull request merged**